// angular-confirmation-popover/esm2015/lib/confirmation-popover.module.js
var ConfirmationPopoverModule_1;
import { __decorate } from "tslib";
import { NgModule, InjectionToken } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ConfirmationPopoverDirective } from './confirmation-popover.directive';
import { ConfirmationPopoverWindowComponent } from './confirmation-popover-window.component';
import { FocusDirective } from './focus.directive';
import { ConfirmationPopoverOptions, } from './confirmation-popover-options.provider';
export const USER_OPTIONS = new InjectionToken('confirmation popover user options');
export function optionsFactory(userOptions) {
    const options = new ConfirmationPopoverOptions();
    Object.assign(options, userOptions);
    return options;
}
let ConfirmationPopoverModule = ConfirmationPopoverModule_1 = class ConfirmationPopoverModule {
    static forRoot(options = {}) {
        return {
            ngModule: ConfirmationPopoverModule_1,
            providers: [
                {
                    provide: USER_OPTIONS,
                    useValue: options,
                },
                {
                    provide: ConfirmationPopoverOptions,
                    useFactory: optionsFactory,
                    deps: [USER_OPTIONS],
                },
            ],
        };
    }
};
ConfirmationPopoverModule = ConfirmationPopoverModule_1 = __decorate([
    NgModule({
        declarations: [
            ConfirmationPopoverDirective,
            ConfirmationPopoverWindowComponent,
            FocusDirective,
        ],
        imports: [CommonModule],
        exports: [ConfirmationPopoverDirective, FocusDirective],
        entryComponents: [ConfirmationPopoverWindowComponent],
    })
], ConfirmationPopoverModule);
export { ConfirmationPopoverModule };
